### PR TITLE
Almost complete rewrite

### PR DESCRIPTION
**Pull request by John-Colvin**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -15,4 +15,5 @@ addons:
       - libopenmpi-dev
 
 script:
-  dub test --build=travis --compiler=${DC}
+ - dub test --compiler=${DC}
```
Comment (9il) [inline .travis.yml:40]: `--build=travis`

**Comment by John-Colvin** [comment]
Still broken, don't merge even if autotester says it's OK

**Comment by 9il** [comment]
How it can be?

**Comment by John-Colvin** [comment]
`mpirun` returns 0 even if it fails :-1:

**Comment by John-Colvin** [comment]
ok, so now everything passes on dmd on the auto-tester, but everything crashes with ldc and gdc. On my archlinux machine only gdc works. On my mac both ldc and dmd work, haven't tested with gdc

**Comment by 9il** [comment]
Is it linking error?

**Comment by John-Colvin** [comment]
I've concluded that the problems go deeper than build configurations in D. I'm doing a full manual review of `mpi/package.d` against the C headers to try and catch the problem.

**Comment by John-Colvin** [comment]
Everything is now correct (as far as i can tell), but unfortunately Travis-CI uses ubuntu 12.04 which has an ancient OpenMPI.

**Comment by John-Colvin** [comment]
Tests are starting to go green for the right reasons!

**Comment by John-Colvin** [comment]
I guess I should explain why I did this:

There were loads of problems with these bindings. Lots of misunderstandings about extern and __gshared, void\* instead of the correct types and no versioning whatsoever. They were also quite out-of-date, Open-MPI has moved on since then. By going through the entire header by hand (almost 18 hours of work in total!), we now have a binding that somewhere near generally usable quality.

A certain amount of configuring is necessary because mpi installations can vary a lot. This is handled in a quite fragile and messy manner in places, but it does actually work on a variety of systems and the end result is not compromised.

**Comment by John-Colvin** [comment]
Hooray, green ticks :+1:

**Comment by John-Colvin** [comment]
P.S. This is ready for merging. Feel free to attempt to review it, but it is rather large. All the actual bindings are in `source/mpi/package.d.in`

**Comment by SFrijters** [comment]
FWIW, I just tried to use this pull request for my code (which was using some hacked MPICH bindings up to this point) and it seems to work, mostly. Would it be possible to add a module level @nogc and nothrow? That is the one thing that was still causing problems for me at the end.

**Comment by John-Colvin** [comment]
@SFrijters Yes, I will do that. I'm working on adding proper MPICH bindings as well, which will appear on the mpi_generic branch when it's ready.

**Comment by John-Colvin** [comment]
@SFrijters P.S. feel free to come join in on https://gitter.im/DlangScience/public, more perspectives will be very useful for designing the overall structure of the project.

**Comment by SFrijters** [comment]
Ok, will take a look at the gitter thing soonish. It would be nice to also get @Laeeth on board - he's been working on new bindings for HDF5 based on my old ones, and as far as I can see it's still locked into my old MPICH-centric types. Once generic MPI bindings are available life will be much easier! Maybe it would even be possible to autogenerate the basic HDF5 bindings similarly to these?

**Comment by John-Colvin** [comment]
@SFrijters Laeeth has been in contact with me about the HDF5 bindings/wrappers, it's on my list of things to look at.

The MPI bindings won't be truly generic, I'm not totally confident how possible that even would be (Perhaps someone with a more intimate knowledge of the MPI spec and common implementations would be able to answer that). What they will be is a bunch of different special cases for the common implementations, i.e. `mpi.h.ompi.in`, `mpi.h.mpich.in`, with auto-detection at build-time of which version is installed (overridable by the user pointing to a specific `mpicc`). There appears to be an effort by MPICH and some other vendors (not including OpenMPI as far as i could see) to provide ABI compatibility, which we should be able to take advantage of, but at the moment it looks like there's quite a lot of variety across implementations.

**Comment by SFrijters** [comment]
@John-Colvin "Quite" generic would already be a big step forward. Ideally I would want to do performance benchmarks of my code on some big iron sometime, which may need Intel/Bullx/Cray MPI or somesuch, but it seems they are based on either MPICH or OpenMPI (Intel claims to be ABI compatible with MPICH), so it may not be too hard to get that to work (getting a proper compiler set up on one of those machines is probably harder...).

**Comment by John-Colvin** [comment]
Unless there are any objections I'm gonna go ahead and merge this soon. It's not perfect, but what we currently have is completely broken so it can't possibly be worse...

**Comment by SFrijters** [comment]
:+1: 

I don't see this on code.dlang.org yet? Could it be tagged and registered?

**Comment by John-Colvin** [comment]
Tagging a release and registering on code.dlang.org will come soon after this is merged.

**Comment by Laeeth** [comment]
Hi Stefan.

Thanks for your message.

I started with yours, which didn't compile for me in beginning, and also had to update as API changed with later versions.  I also ported quite a few of test programs and started writing more idiomatic D wrapper.  Very simple so far.

Aldacron is also working on bindings, which I haven't yet looked at.  They don't seem to be so frequently updated.  He is more experienced with D than me, but I have a different idea philosophically to him.  He prefers classes to structs, but I don't think that reproducing the Byzantine class structure makes sense for D.

So better in my view to have:
- simple binding as declaration
- replication of C API with nicer interface so you can pass a string not a char, ubyte[], etc and use exceptions not error codes
- idiomatic D style wrapper that deals with type conversion and so on.

For idiomatic D wrapper, either you know beforehand what the data is and just check it fits.  So you try and read a range of structs from the file and it enforces correctness.  Or you don't know, in which case you need to have a dataframe.  A dataframe is just a matrix of variants where all columns are the same type and they have titles.  I have written a simple dataframe and Vlad Levenfeld is helping me.

You can have a look at my github dataframe library to see.  It's not clever or pretty, but it's a start and works for me.

When you have a dataframe it's easy to open up any 2d h5 file.  Simple idea but very useful.

It's not hard to make it compatible with pandas, even so it is auto converted using PyD.

Two limiting factors are time and my lack of full knowledge of HDF5.  I don't use MPI, so someone needs to help with that.  And my career is to be an investor, not programmer, so I built this because I needed it.  Going forward I will have much less time.  I may be paying someone part time in D community to help continue my work, but he will have a lot on his plate too, so I welcome any help.

Mine work for me for what I want to do, but if someone would like to be co-maintainer or even take over primary responsibility that would be great.

I think objectively speaking D is great for middling data, but the libraries are missing.  And HDF5 must be a key part of that as it's by far the most sensible data format for a broad range of cases.

We need to convert the sample programs into unit tests.

I broke Windows support, so that needs to be put back too sometime.  And no idea about OSX.

Laeeth.

**Comment by SFrijters** [comment]
@Laeeth Let's continue the HDF5 discussion someplace else; I'll raise an issue in your repo.

**Comment by Laeeth** [comment]
Tku.  Sounds good.  L

**Comment by drug007** [comment]
I'd like to discuss the HDF5 too, where can it be done?

**Comment by Laeeth** [comment]
please open an issue here @drug007  https://github.com/Laeeth/d_hdf5
and ready for whenever suits @SFrijters

**Pull request merged**